source/_posts/2025-06-25-piping-in-new-features.md

```diff
@@ -0,0 +1,164 @@
+
+Sadly, PFA didn't pass due to some engine complexity issues, and that largely undermined the v2 Pipe RFC, too.  However, we did get a consolation prize out of it: [First Class Callables](https://wiki.php.net/rfc/first_class_callable_syntax) (the `array_values(...)` syntax), courtesy Nikita Popov, were by design a "junior", degenerate version of partial function application.
+
+Fast forward to 2025, and I was sufficiently bored to take another swing at pipes.  This time with a better implementation with lots of hand-holding from Ilija Tovilo and Arnaud Le Blanc, both part of the PHP Foundation dev team, I was able to get it through.
+
+Third time's the charm.
+
+## The implications
+
+Above, we described pipes as "deceptively simple."  The implementation itself is almost trivial; it's just syntax sugar for the temp variable version, effectively.  However, the best features are the ones that can combine with others or be used in novel ways to punch above their weight.
+
+We saw above how a long array manipulation process could now be condensed into a single chained expression.  Now imagine using that in places where only a single expression is allowed, such as a `match()`:
+
+```php
+$string = 'something GoesHERE';
+
+$newString = match ($format) {
+    'snake_case' => $string
+        |> splitString(...)
+        |> fn($x) => implode('_', $x)
+        |> strtolower(...),
+    'lowerCamel' => $string
+        |> splitString(...),
+        |> fn($x) => array_map(ucfirst(...), $x)
+        |> fn($x) => implode('', $x)
+        |> lcfirst(...),
+    // Other case options here.
+};
+```
+
+Or, consider that the right-side can also be a function call that returns a `Closure`.  That means with a few functions that return functions:
+
+```php
+$profit = [1, 4, 5] 
+    |> loadSeveral(...)
+    |> filter(isOnSale(...))
+    |> map(sellWidget(...))
+    |> array_sum(...);
+```
+
+Which... gives us mostly the same thing as the long-discussed scalar methods!  Only pipes are more flexible as you can use any function on the right-side, not just those that have been blessed by the language designers as methods.
+
+At this point, pipe comes very close to being "extension functions", a feature of Kotlin and C# that allows writing functions that look like methods on an object, but are actually just stand-alone functions.  It's spelled a bit differently (`|` instead of `-`), but it's 75% of the way there, for free.
+
```

Comment:
It would be nice to add a subheader here if possible. Wdyt?

Reply:
Eh, any one of these is somewhat arbitrary.  They're an increasing scale of cool implications.  I'd say put a 3rd level header on all of them, or on none of them.  I lean none, but can do either.